### PR TITLE
#737 Ensure types given for @Mapper#uses() and config() are valid

### PR DESCRIPTION
**Pull request by Captain1653**
@sjaakd  and @filiphr I need you help with issue #737 .
1. Do we still need to do it? :) (It's a little old issue)
2. I have some implementation (sure, it isn't so good), but there are some strange behavior. 
Tests for module processor are green. Integration tests are failed in class FullFeatureCompilationTest. My tests in idea failed with eclipse. (Is it some specific thing with eclipse?)
3. In class CompilationStatement I have 9 identical errors (honestly, I don't know. that's why I use discinct. How can I avoid it? like `filterExpectedDiagnostics` or something like that ? Am I doing something wrong?
4. Can you say me right direction or give some advice? Should I leave this issue, because it isn't my level and too complicated for me?

full build
tests module processor
failed test and filter

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java [hunk 1]
```diff
@@ -42,19 +43,29 @@
...
+    private MapperConfiguration(MapperPrism mapperPrism, FormattingMessager messager) {
         this.mapperPrism = mapperPrism;
+        this.messager = messager;
 
         if ( mapperPrism.values.config() != null ) {
-            // TODO #737 Only a declared type makes sense here; Validate and raise graceful error;
-            // Also validate that @MapperConfig is present
-            this.config = (DeclaredType) mapperPrism.config();
-            this.mapperConfigPrism = MapperConfigPrism.getInstanceOn( config.asElement() );
+            if ( mapperPrism.config() instanceof DeclaredType ) {
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java]: I think the config itself even needs to be an interface.. so you can check that as well

File: processor/src/test/java/org/mapstruct/ap/test/mypackage/CheckDeclaredTypeInMapperUses.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                kind = javax.tools.Diagnostic.Kind.ERROR,
+                line = 16,
+                messageRegExp = "Type \"int\" cannot be used in \"@Mapper\\.uses\\(\\)\""
+            )
+        }
+    )
+    public void shouldCompilationFailedForMapperUsesNotDeclaredType() { }
+
+    @Test
+    @WithClasses( { MapperWithNotDeclaredTypeInMapperConfigUses.class, MapperConfigWithNotDecladerTypeInUses.class} )
```
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/mypackage/CheckDeclaredTypeInMapperUses.java]: There's a convention: mappers that give errors should start with `ErroneousXYZ`..  Perhaps that's the problem in your full feature integration test

File: processor/src/main/java/org/mapstruct/ap/internal/model/AbstractBaseBuilder.java [hunk 3]
```diff
@@ -60,7 +60,8 @@ boolean canGenerateAutoSubMappingBetween(Type sourceType, Type targetType) {
     }
 
     private boolean isDisableSubMappingMethodsGeneration() {
-        MapperConfiguration configuration = MapperConfiguration.getInstanceOn( ctx.getMapperTypeElement() );
+        MapperConfiguration configuration = MapperConfiguration.getInstanceOn(
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/AbstractBaseBuilder.java]: I'm not so happy with the amount of `getInstanceOn` that we do and the distributed checking of the config. Shouldn't we treat it more like we treat other annotations: have a class doing the checking in our `org.mapstruct.ap.internal.model.source` package?

**Comment by sjaakd** [comment]
> 1. Do we still need to do it? :) (It's a little old issue)

I think so :).

**Comment by sjaakd** [comment]
> 2\. I have some implementation (sure, it isn't so good), but there are some strange behavior.
>     Tests for module processor are green. Integration tests are failed in class FullFeatureCompilationTest. My tests in idea failed with eclipse. (Is it some specific thing with eclipse?)

Don't worry.. 

Could it be the erroneous mapper name? just guessing here.

**Comment by sjaakd** [comment]
> 3\. n class CompilationStatement I have 9 identical errors (honestly, I don't know. that's why I use discinct. How can I avoid it? like `filterExpectedDiagnostics` or something like that ? Am I doing something wrong?

I'm not sure here .. are these coming from the same test case?

**Comment by Captain1653** [comment]
> > 3. n class CompilationStatement I have 9 identical errors (honestly, I don't know. that's why I use discinct. How can I avoid it? like `filterExpectedDiagnostics` or something like that ? Am I doing something wrong?
> 
> I'm not sure here .. are these coming from the same test case?
Yes, from one test case. As you can see we use `getInstanceOn` many times that's why we receive so many identical warnings/errors. When I'll check config in one place this problem with be solved.

**Comment by Captain1653** [comment]
> > 3. n class CompilationStatement I have 9 identical errors (honestly, I don't know. that's why I use discinct. How can I avoid it? like `filterExpectedDiagnostics` or something like that ? Am I doing something wrong?
> 
> I'm not sure here .. are these coming from the same test case?

Yes, from one test case. As you can see we use getInstanceOn many times that's why we receive so many identical warnings/errors. When I'll check config in one place this problem with be solved.

File: processor/src/main/java/org/mapstruct/ap/internal/model/AbstractBaseBuilder.java (hunk 3 above)
Comment (Captain1653) [reply processor/src/main/java/org/mapstruct/ap/internal/model/AbstractBaseBuilder.java]: Maybe I'm wrong, but unfortunatelly no. We haven't class in org.mapstruct.ap.internal.model.source for mapper configuration. We always use MapperConfiguration.getInstanceOn eight times. Sometimes I think, it's too early to do this task with checking. Firstly we should refactor, maybe. What do you think?

**Comment by Captain1653** [comment]
I think right now it's too earlt for this PR. Firstly we should refactor our code and create MapperConfig in source package, like said @sjaakd . I close it.

**Comment by sjaakd** [comment]
@Captain1653 .. thanks.

**Pull request closed without merging**